pennylane/transforms/qfunc_transforms.py

```diff
@@ -0,0 +1,386 @@
+                qml.RX(0.1, wires=0)
+
+                # apply the transform to the ansatz
+                my_transform(*transform_weights)(ansatz)(param)
+
+                return qml.expval(qml.PauliZ(1))
+
+        We can print this QNode to show that the qfunc transform is taking place:
+
+        >>> x = np.array(0.5, requires_grad=True)
+        >>> y = np.array([0.1, 0.2], requires_grad=True)
+        >>> print(qml.draw(circuit)(x, y))
+         0: ──RX(0.1)───H──────────╭Z──┤
+         1: ──RX(0.05)──RY(0.141)──╰C──┤ ⟨Z⟩
+
+        Evaluating the QNode, as well as the derivative, with respect to the gate
+        parameter *and* the transform weights:
+
+        >>> circuit(x, y)
+        0.9887793925354269
+        >>> qml.grad(circuit)(x, y)
+        (array(-0.02485651), array([-0.02474011, -0.09954244]))
+
+        **Implementation details**
+
+        Internally, the qfunc transform works as follows:
+
+        .. code-block:: python
+
+            def transform(old_qfunc, params):
+                def new_qfunc(*args, **kwargs):
+                    # 1. extract the tape from the old qfunc, being
+                    # careful *not* to have it queued.
+                    tape = make_tape(old_qfunc)(*args, **kwargs)
+
+                    # 2. transform the tape
+                    new_tape = tape_transform(tape, params)
+
+                    # 3. queue the *new* tape to the active queuing context
+                    new_tape.queue()
+                return new_qfunc
+
+        *Note: this is pseudocode; the actual implementation is significantly more complicated!*
+
+        Steps (1) and (3) are identical for all qfunc transforms; it is only step (2),
+        ``tape_transform`` and the corresponding tape transform parameters, that define the qfunc
+        transformation.
+
+        That is, given a tape transform that **defines the qfunc transformation**, the
+        decorator **elevates** the tape transform to one that works on quantum functions
+        rather than tapes. This decorator therefore automates the process of adding in
+        the queueing logic required under steps (1) and (3), so that it does not need to be
+        repeated and tested for every new qfunc transform.
+    """
+    if not callable(tape_transform):
+        raise ValueError(
+            "The qfunc_transform decorator can only be applied "
+            "to single tape transform functions."
+        )
+
+    if not isinstance(tape_transform, single_tape_transform):
+        tape_transform = single_tape_transform(tape_transform)
+
+    sig = inspect.signature(tape_transform)
+    params = sig.parameters
+
+    if len(params) > 1:
+
+        @functools.wraps(tape_transform)
+        def make_qfunc_transform(*targs, **tkwargs):
+            def wrapper(fn):
+                return _create_qfunc_internal_wrapper(fn, tape_transform, targs, tkwargs)
+
+            return wrapper
+
+    elif len(params) == 1:

```

Comment:
Why not `else`?

Should you raise an error if `len(params)==0`?

Reply:
I don't think `len(params)` can be zero here; this invalid edge case would have already been caught by `single_tape_transform(tape_transform)`